File: Astro-TFG/public/scripts/3-carga-graficos-complejos/prueba-2/version-1/python/main.py
import time
import json
import asyncio
from pyscript import PyWorker, display
import js  # type: ignore

worker = None


async def initialize_worker():
    global worker
    if worker is None:
        origin = js.window.location.origin
        py_path = js.window.document.body.dataset.pyPath
        worker = PyWorker(
            f"{origin}{py_path}worker.py",
            type="pyodide",
            config=f"{origin}{py_path.replace('python', 'json')}pyscript-worker.json")
        await worker.ready


async def launch_worker(event):
    global worker, worker_time

    start_w = time.perf_counter()
    await initialize_worker()
    worker_time = (time.perf_counter() - start_w) * 1000

    result_json = await worker.sync.do_analisis()
    result = json.loads(result_json)

    graph_json_str = json.dumps(result["graph"])

    js.displayPlotFromJSON(graph_json_str, "pyscript-output")
    js.window.startFPSMeasurement(3000, "pyscript-output")

    js.window.hideExecutionLoader()
    display_result(result=result["metrics"])


def display_result(result):
    display(f"Worker init time: {worker_time:.2f} ms",
            target="pyscript-output")
    display(
        f"Data generation: {result['data_gen_time']:.2f} ms", target="pyscript-output")
    display(
        f"Rendering: {result['render_time']:.2f} ms", target="pyscript-output")
    display(f"Memory: {result['memory']:.2f} MB", target="pyscript-output")
    display(
        f"Total ET: {result['total_time_ms']:.2f} ms", target="pyscript-output")


def js_run_py_benchmark(event):
    js.clearCell('pyscript-output')
    asyncio.ensure_future(launch_worker(None))


js.window.run_py_benchmark = js_run_py_benchmark

File: Astro-TFG/public/scripts/3-carga-graficos-complejos/prueba-2/version-1/javascript/main.js
let worker = null;
let workerTime = 0;

function initializeWorker() {
    if (!worker) {
        worker = new Worker(
            window.location.origin +
            window.document.body.dataset.jsPath +
            "worker.js"
        );
    }
}

async function _runJsBenchmark() {
    try {
        window.clearCell("javascript-output");
        window.showExecutionLoader();

        const start_time_worker = performance.now();
        initializeWorker();
        workerTime = performance.now() - start_time_worker;

        const num_points = parseInt(
            document.getElementById("num-points-javascript").value,
            10
        );
        const num_series = parseInt(
            document.getElementById("num-series-javascript").value,
            10
        );
        const id = `js-${Date.now()}`;

        const resultJson = await new Promise((resolve, reject) => {
            function onMessage(e) {
                if (e.data.id !== id) return;
                worker.removeEventListener("message", onMessage);

                if (e.data.error) {
                    reject(new Error(e.data.error));
                }
                else if (e.data.payload !== undefined) {
                    const payload = {
                        metrics: e.data.payload.metrics,
                        data: e.data.payload.graphData.traces,
                        layout: e.data.payload.graphData.layout
                    };
                    resolve(JSON.stringify(payload));
                }
                else {
                    reject(new Error("Worker dont return results"));
                }
            }

            worker.addEventListener("message", onMessage);
            worker.postMessage({
                id,
                type: "do_analisis",
                num_points,
                num_series,
            });
        });

        const result = JSON.parse(resultJson);
        displayResult(result.metrics);

        window.displayPlotFromJSON(resultJson, 'javascript-output');
        window.startFPSMeasurement(3000, "javascript-output");

    } catch (err) {
        console.error("Error:", err);
    } finally {
        window.hideExecutionLoader();
    }
}

function createDiv() {
    const div = document.createElement("div");
    return div;
}

function displayResult(r) {
    const output = document.getElementById("javascript-output");

    const workerDiv = createDiv();
    workerDiv.textContent = `Worker init time: ${workerTime.toFixed(2)} ms`;
    output.appendChild(workerDiv);

    const avgDataGenDiv = createDiv();
    avgDataGenDiv.textContent = `Data generation: ${r.dataGenTime.toFixed(2)} ms`;
    output.appendChild(avgDataGenDiv);

    const avgRenderDiv = createDiv();
    avgRenderDiv.textContent = `Rendering: ${r.renderTime.toFixed(2)} ms`;
    output.appendChild(avgRenderDiv);

    const memoryDiv = createDiv();
    memoryDiv.textContent = `Memory: ${r.mem.toFixed(2)} MB`;
    output.appendChild(memoryDiv);

    const totalTimeDiv = createDiv();
    totalTimeDiv.textContent = `Total ET: ${r.totalTime.toFixed(2)} ms`;
    output.appendChild(totalTimeDiv);
}

window.runJsBenchmark = _runJsBenchmark;
